Editor: Hide SEO description for private/hidden sites

jeremeylduvall opened the pull request:
We disable the SEO settings for any site that is private. This commit disables the SEO description drawer in the Editor when a site is private as well. This is the second attempt after #11476 broke production (see p4TIVU-5UZ-p2).

The previous attempt broke production because it assumed `this.props.siteSettings` would exist when someone tried to load the editor. With a clear cache, this isn't necessarily true. We fetch site settings in specific instances like when someone tries to [visit the settings page directly](https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/site-settings/controller.js#L85). When users tried to load the editor with an empty initial state, they received an error because `blog_public` wasn't defined yet leading to a broken experience.

In the revised approach:
- We use the specific `getSiteSetting` selector under [`state/selectors/`](https://github.com/Automattic/wp-calypso/tree/master/client/state/selectors).
- There are two safety valves. First, we check the truthiness of `this.props.privacySetting`. Then, we set a default value of `1` as a fallback.

I did test this on a cleared cache in my regular browser and an incognito window. I didn't experience any of the errors from the previous attempt.

## To test
1. Open up your site settings for a site on the Business plan: http://calypso.localhost:3000/settings/general/
2. Change the site to public
3. Click "Add" next to posts to start a new post. Verify that you see the SEO description panel in the editor drawer.
4. Click "My Sites" and go back to your site settings. Set the site to either hidden or private.
5. Start a new post again and verify that you do not see the SEO description panel in the drawer.

Resolves: #11406

aduth commented on client/post-editor/editor-drawer/index.jsx:
```diff
@@ -246,7 +247,10 @@ const EditorDrawer = React.createClass( {
 		const { plan } = this.props.site;
 		const hasBusinessPlan = isBusiness( plan ) || isEnterprise( plan );
-		if ( ! hasBusinessPlan ) {
+		const { privacySetting = 1 } = this.props;
```
It doesn't appear that the default value here would ever be used; partly because the condition above `! this.props.privacySetting` would abort on a falsey value, but also because `mapStateToProps` always provides at least a `null` value based on the behavior of `getSiteSetting`. This would not trigger the default ([see example](http://jsbin.com/pecomacahu/edit?js,console)).

jeremeylduvall commented:
@aduth Thanks for the review. I'm now rendering in the `QuerySiteSettings` data component. In my testing, selectors that depend on the site settings object specifically versus the site object are more responsive to changes. With the site object, if I update the privacy options, there have been instances where `isPrivateSite` returns the incorrect value. I'm assuming that's because they're from different endpoints?

aduth commented:
>With the site object, if I update the privacy options, there have been instances where isPrivateSite returns the incorrect value. I'm assuming that's because they're from different endpoints?

Yeah, based on my understanding, this is expected but could be improved. I ran into a similar case with site icon, in that it's duplicated between both the settings endpoint and the sites endpoint. In my case, I deferred to having the sites state be the preferred (canonical) source, tracking updates to site settings in the sites reducer. We should probably do something similar for privacy settings.

- https://github.com/Automattic/wp-calypso/pull/10016
- https://github.com/Automattic/wp-calypso/blob/3e6771f/client/state/selectors/get-site-icon-id.js#L21-L32
- https://github.com/Automattic/wp-calypso/blob/3e6771f/client/state/sites/reducer.js#L135-L175

jeremeylduvall commented:
Good call. I like that approach.

I modified the sites reducer to monitor changes in blog privacy as well. I left comments to this effect in the code for clarity, but since the settings endpoint returns a numerical value for `blog_public`, we need to use the `parseInt` expression to convert that to true/false. The `isPrivateSite` selector now first looks at the site state and has a fallback to site settings.

Extra sanity check since it broke prod last time, with the updates to `isPrivateSite`, a check in [`editor-drawer/index.jsx`](https://github.com/Automattic/wp-calypso/blob/93af5bbc2a1ff70f9888d9509ce4f920637bcf2d/client/post-editor/editor-drawer/index.jsx#L238) is no longer necessary to prevent errors on a clear cache. I tested on a cleared cache in an incognito window. Worked fine on my side.

Tests for `isPrivateSite` are now failing. Will update now.

jeremeylduvall commented:
Tests for `isPrivateSite` are now updated to match the selector behavior. All tests are passing locally. You can run the specific test for `isPrivateSite` using the following command:

`npm run test-client client/state/selectors/test/is-private-site.js`

jeremeylduvall commented:
Tests should all be passing now. Long story (hopefully) short: With the updated behavior in `isSitePrivate`,  we now [fall back](https://github.com/Automattic/wp-calypso/blob/7d08e9cdc7e98ce7cda95d3b8e5aa524f99c84db/client/state/selectors/is-private-site.js#L23) on site settings if site doesn't exist. This calls [`getSiteSettings`](https://github.com/Automattic/wp-calypso/blob/f0cedac580d2239a73a8b463cd59a2403eca8b3c/client/state/site-settings/selectors.js#L46).

We use `isPrivateSite` in [`isSiteSupportingImageEditor`](https://github.com/Automattic/wp-calypso/blob/7d08e9cdc7e98ce7cda95d3b8e5aa524f99c84db/client/state/selectors/is-site-supporting-image-editor.js#L17). In the test file, when you provide an empty `sites` object, you also need to provide an empty `siteSettings` object. Otherwise, `getSiteSettings` will fail on the fall back. I've added [a comment](https://github.com/Automattic/wp-calypso/blob/7d08e9cdc7e98ce7cda95d3b8e5aa524f99c84db/client/state/selectors/test/is-site-supporting-image-editor.js#L17) to this effect for clarity, but I can remove if it's not necessary.

aduth commented on client/post-editor/editor-drawer/index.jsx:
```diff
@@ -328,7 +331,8 @@ export default connect(
 			jetpackVersionSupportsSeo: isJetpackMinimumVersion( state, siteId, '4.4-beta1' ),
-			typeObject: getPostType( state, siteId, type )
+			typeObject: getPostType( state, siteId, type ),
+			isSitePrivate: isPrivateSite( state, siteId ),
```
Doing cartwheels to avoid [`no-shadow`](http://eslint.org/docs/rules/no-shadow) ? 😛 Seems fine as `isPrivate` too if that would be clearer than the word shuffling.

jeremeylduvall replied:
Ha I take ESLint rules very seriously :)

aduth commented on client/state/selectors/is-private-site.js:
```diff
@@ -12,9 +13,17 @@ import { getRawSite } from 'state/sites/selectors';
+
+	if ( settings ) {
+		// Site settings returns a numerical value for blog_public.
+		return parseInt( settings.blog_public, 10 ) !== 1;
```
Can we rely on `settings.blog_public` to always be a string? My hunch is maybe not (via optimistic state updates), in which case what you have here is probably good for safety. If we could be confident, we might be able to simply `return '1' !== settings.blog_public`.

jeremeylduvall replied:
I'd prefer to keep the current behavior. I feel like I've seen multiple behaviors while testing this out.

aduth commented on client/state/sites/reducer.js:
```diff
@@ -138,34 +138,49 @@ export function items( state = {}, action ) {
-					}
+					is_private: sitePrivacy
 				};
+			} else if ( settings.hasOwnProperty( 'site_icon' ) ) {
```
Couldn't a `settings` payload have both `blog_public` _and_ `site_icon`? In which case we'd only update one or the other because of `else if`. Probably better as a separate `if`.

aduth replied:
>Couldn't a `settings` payload have both `blog_public` _and_ `site_icon`? In which case we'd only update one or the other because of `else if`. Probably better as a separate `if`.

To this point, we should probably add a test for this case.

jeremeylduvall replied:
> Couldn't a settings payload have both blog_public and site_icon? In which case we'd only update one or the other because of else if. Probably better as a separate if.

I ran through some scenarios, and I don't think the settings payload will actually ever contain both a `blog_public` and `site_icon` update. When the site icon is updated, we call [`saveSiteSettings` immediately](https://github.com/Automattic/wp-calypso/blob/3ed0e7d0a658ca7656f190401fa82996885bcefe/client/my-sites/site-settings/site-icon-setting/index.jsx#L93) which automatically saves the settings with only the icon property. You can see this if you change the blog privacy (but don't save), update the icon (see the "Settings Saved" dialogue), and then try to leave the page. You'll see an error because the blog privacy setting is still queued.

Happy to update it either way just in case this behavior is changed in the future.

aduth replied:
I think the more likely scenario is that we receive an entire fresh set of site settings via `SITE_SETTINGS_RECEIVE`, which would always include both keys, one or the other or both potentially having been updated.

aduth commented on client/state/sites/reducer.js:
```diff
@@ -138,34 +138,49 @@ export function items( state = {}, action ) {
+			if ( settings.hasOwnProperty( 'blog_public' ) ) {
+				// Site settings returns a numerical value for blog_public.
+				// Site state expects a true/false value for is_private.
+				const sitePrivacy = parseInt( settings.blog_public, 10 ) !== 1;
```
Minor: Might be nice to indicate boolean value by prefixing with `is`: `const isPrivate = ...`

aduth commented on client/state/sites/reducer.js:
```diff
@@ -138,34 +138,49 @@ export function items( state = {}, action ) {
 			// A site settings update may or may not include the icon property.
 			// If not, we should simply return state unchanged.
-			if ( ! settings.hasOwnProperty( 'site_icon' ) ) {
+			if ( ! settings.hasOwnProperty( 'site_icon' ) && ! settings.hasOwnProperty( 'blog_public' ) ) {
```
Not sure what gotchas you might encounter, but doing a `reduce` on the keys we're monitoring might be a nicer way to encapsulate the separate logic.

Something like:

```js
return reduce( [ 'blog_public', 'site_icon' ], ( memo, key ) => {
	if ( ! settings.hasOwnProperty( key ) ) {
		return memo;
	}

	switch ( key ) {
		case 'blog_public':
			// ...

		case 'site_icon':
			// ...
	}

	return memo;
}, state );
```

(`SITE_RECEIVE` reducer above is similar in how it reduces using state as initial value while avoiding direct mutations)

jeremeylduvall commented on client/state/sites/test/reducer.js:
```diff
@@ -412,7 +412,7 @@ describe( 'reducer', () => {
 			} );
 
-			expect( state ).to.equal( original );
+			expect( state ).to.eql( original );
```
Since we're now returning memo, which is a copy of state, it was necessary to change these to `eql`.

aduth replied:
That we needed to change the tests here implies that we're returning a new state reference in the reducer where we shouldn't be (no changes occurred). I think in your reducer's `switch` where you're breaking upon finding that there are no differences, we should `return memo;` instead to bypass the rest of the `reduce` and ensure that we only create a copy of the state object (return a new reference) if we're certain that a change has actually occurred.

jeremeylduvall replied:
Good call. Updated.

jeremeylduvall commented:
Updated the sites reducer to use a switch statement, which helped to clear up the code a bit. It's also easier to add additional settings in case we want to repeat this behavior in the future. Everything should still be working as expected.

aduth commented on client/state/sites/reducer.js:
```diff
@@ -135,43 +135,58 @@ export function items( state = {}, action ) {
+						if ( null === mediaId ) {
+							// Unset icon
+							nextSite = omit( nextSite, 'icon' );
+							break;
```
This `break;` is redundant since there's no other logic in the `case` to be run except the final `break;` anyways.

aduth commented on client/state/sites/reducer.js:
```diff
@@ -135,43 +135,58 @@ export function items( state = {}, action ) {
+				switch ( key ) {
+					case 'blog_public':
+						const isPrivate = parseInt( settings.blog_public, 10 ) !== 1;
+						nextSite.is_private = isPrivate;
```
I think we'll want to check (and add test case for) whether the value has actually changed. In this case we're always updating the site state even if the value we've received is the same as what we're currently tracking.

aduth commented on client/state/sites/reducer.js:
```diff
@@ -135,43 +135,58 @@ export function items( state = {}, action ) {
 			}
 
-			const mediaId = settings.site_icon;
+			let nextSite = { ...site };
```
Minor-ish: We're creating a shallow clone of the site even if we never end up making any changes. We could do this at the point we need it instead:

```js
let nextSite = site;

// ...

case 'blog_public':
	const isPrivate = parseInt( settings.blog_public, 10 ) !== 1;
	if ( site.is_private === isPrivate ) {
		return memo;
	}

	nextSite = { ...nextSite, is_private: isPrivate };
	break;
```

Only downside of this is that if multiple keys have new values, we'd clone the site multiple times. I think it's the much more likely scenario that nothing will have changed (thus zero cloning) than multiple changed values though.

jeremeylduvall commented:
I went back and forth on whether or not to clone `site` outside of the `switch` block so I could avoid doing the repeated `nextSite` gymnastics.
```js 
nextSite = { ...nextSite, // ... }
```
You make a good point about frequency. I updated the behavior and added a test case for:

> should return same state if received privacy setting and matches current value

jeremeylduvall commented:
I'm not seeing the SEO description field on my side when a site is changed to hidden. `site.is_private` should return `true` if the `blog_public` setting is set to `-1` (private) or `0` (hidden). It only returns `false` if `blog_public` is set to `1` (public).

Here's a GIF of what I'm experiencing with a hidden site:

![private](https://cloud.githubusercontent.com/assets/7240478/23756175/10902430-04a0-11e7-8905-44425936abcc.gif)

aduth commented:
The problem is that the reducer logic for `is_private` is not accurate.

In this pull request, we test:

```js
const isPrivate = parseInt( settings.blog_public, 10 ) !== 1;
```

Server-side, we test:

```php
return ( -1 == get_option( 'blog_public' ) );
```

Key difference being that our reducer is treating `blog_public` of `0` as a private site, but this is not true as considered server-side.

jeremeylduvall commented:
Per a Slack convo with @aduth, I updated this PR to treat `-1` as private and `0` (hidden) as public to match the logic we use on the server. This means that the SEO description will still display for hidden sites at the moment. I'll fix this by introducing a new selector for `isHiddenSite`. The logic for the SEO description would then be:

```js
if ( ! hasBusinessPlan || isPrivate || isHidden ) {
  	return;
}
```

CircleCI tests are failing. I'll rebase after #12046, and rerun the tests.

lancewillett commented:
Really happy to see this one. Nice work! 🎉
